Lib/idlelib/sidebar.py

```diff
@@ -297,20 +307,206 @@ def update_sidebar_text(self, end):
+
+        self.delegate.insert(index, chars, tags)
+
+        if is_single_line:
+            after_displaylines = get_displaylines(self, index)
+            if after_displaylines == before_displaylines:
+                return  # no need to update the sidebar
+
+        self.callback()
+
+    def delete(self, index1, index2=None):
+        if index2 is None:
+            index2 = index1 + "+1c"
+        is_single_line = get_lineno(self, index1) == get_lineno(self, index2)
+        if is_single_line:
+            before_displaylines = get_displaylines(self, index1)
+
+        self.delegate.delete(index1, index2)
+
+        if is_single_line:
+            after_displaylines = get_displaylines(self, index1)
+            if after_displaylines == before_displaylines:
+                return  # no need to update the sidebar
+
+        self.callback()
+
+
+class ShellSidebar:
+    """Sidebar for the PyShell window, for prompts etc."""
+    def __init__(self, editwin):
+        self.editwin = editwin
+        self.parent = editwin.text_frame
+        self.text = editwin.text
+
+        self.canvas = tk.Canvas(self.parent, width=30,
+                                borderwidth=0, highlightthickness=0,
+                                takefocus=False)
+
+        self.bind_events()
+
+        change_delegator = \
+            WrappedLineHeightChangeDelegator(self.change_callback)
+
+        # Insert the TextChangeDelegator after the last delegator, so that
+        # the sidebar reflects final changes to the text widget contents.
+        d = self.editwin.per.top
+        if d.delegate is not self.text:
+            while d.delegate is not self.editwin.per.bottom:
+                d = d.delegate
+        self.editwin.per.insertfilterafter(change_delegator, after=d)
+
+        self.text['yscrollcommand'] = self.yscroll_event
+
+        self.is_shown = False
+
+        self.update_font()
+        self.update_colors()
+        self.update_sidebar()
+        self.canvas.grid(row=1, column=0, sticky=tk.NSEW, padx=2, pady=0)
+        self.is_shown = True
+
+    def change_callback(self):
+        if self.is_shown:
+            self.update_sidebar()
+
+    def update_sidebar(self):
+        text = self.text
+        text_tagnames = text.tag_names
+        canvas = self.canvas
+
+        canvas.delete(tk.ALL)
+
+        index = text.index("@0,0")
+        while True:
+            lineinfo = text.dlineinfo(index)
+            if lineinfo is None:
+                break
+            y = lineinfo[1]
+            prev_newline_tagnames = text_tagnames(f"{index} linestart -1c")
+            prompt = (
+                '>>>' if "console" in prev_newline_tagnames else
+                '...' if "stdin" in prev_newline_tagnames else
+                '   '
+            )
+            canvas.create_text(2, y, anchor=tk.NW, text=prompt,
+                               font=self.font, fill=self.colors[0])
+            index = text.index(f'{index}+1line')
```

Comment:
```suggestion
            index = text.index(f'{index}+1line linestart')
```
Sometimes the `+1line` will point to the end of a wrapped line whereas the prompt should be anchored to the linestart

Reply:
According to [the Tk 8.6 docs](http://www.tcl.tk/man/tcl8.6/TkCmd/text.htm#M37), unless we use "+1 display lines", this treats each wrapped line as a single line (that's why this code uses this mechanism!).

Also according to the same docs, it will point to the end of the line of `'{index}'` only if there is no following line, but IDLE makes sure to always keep an extra '\n' and the end of the text widget's contents, so that should never happen.

The only case where this can happen is at the beginning, when the line at `@0,0` (the top of the window) is the end of a wrapped line. In that case we should immediately skip to the beginning of the next line before entering the loop. I've added a fix and a test for that just now.

Thanks for bringing this up!!